src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToComponentCodeActionProvider.cs

```diff
@@ -0,0 +1,214 @@
+        MarkupElementSyntax endNode,
+        string @namespace)
+    {
+        var selectionSpan = AreSiblings(startNode, endNode)
+            ? TextSpanFromNodes(startNode, endNode)
+            : GetEncompassingTextSpan(startNode, endNode);
+
+        return new ExtractToComponentCodeActionParams
+        {
+            Uri = context.Request.TextDocument.Uri,
+            Start = selectionSpan.Start,
+            End = selectionSpan.End,
+            Namespace = @namespace
+        };
+    }
+
+    private static TextSpan GetEncompassingTextSpan(MarkupElementSyntax startNode, MarkupElementSyntax endNode)
+    {
+        // Find a valid node that encompasses both the start and the end to
+        // become the selection.
+        SyntaxNode commonAncestor = endNode.Span.Contains(startNode.Span)
+            ? endNode
+            : startNode;
+
+        while (commonAncestor is MarkupElementSyntax or
+                MarkupTagHelperAttributeSyntax or
+                MarkupBlockSyntax)
+        {
+            if (commonAncestor.Span.Contains(startNode.Span) &&
+                commonAncestor.Span.Contains(endNode.Span))
+            {
+                break;
+            }
+
+            commonAncestor = commonAncestor.Parent;
+        }
+
+        // If walking up the tree was required then make sure to reduce
+        // selection back down to minimal nodes needed.
+        // For example:
+        //   <div>
+        //     {|result:<span>
+        //      {|selection:<p>Some text</p>
+        //     </span>
+        //     <span>
+        //       <p>More text</p>
+        //     </span>
+        //     <span>
+        //     </span>|}|}
+        //   </div>
+        if (commonAncestor != startNode &&
+            commonAncestor != endNode)
+        {
+            SyntaxNode? modifiedStart = null, modifiedEnd = null;
+            foreach (var child in commonAncestor.ChildNodes().Where(static node => node.Kind == SyntaxKind.MarkupElement))
+            {
+                if (child.Span.Contains(startNode.Span))
+                {
+                    modifiedStart = child;
+                    if (modifiedEnd is not null)
+                        break; // Exit if we've found both
+                }
+
+                if (child.Span.Contains(endNode.Span))
+                {
+                    modifiedEnd = child;
+                    if (modifiedStart is not null)
+                        break; // Exit if we've found both
+                }
+            }
+
+            if (modifiedStart is not null && modifiedEnd is not null)
+            {
+                return TextSpanFromNodes(modifiedStart, modifiedEnd);
+            }
+        }
+
+        // Fallback to extracting the nearest common ancestor span
+        return commonAncestor.Span;
+    }
+
+    private static TextSpan TextSpanFromNodes(SyntaxNode start, SyntaxNode end)
+        => new(start.Span.Start, end.Span.End - start.Span.Start);
```

Comment:
Nit: Could remove this method, as `TextSpan.FromBounds(start.Span.Start, end.Span.End)` does the same thing